File: app/src/main/java/com/mkulesh/onpc/iscp/scripts/AutoPower.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2024 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */
package com.mkulesh.onpc.iscp.scripts;

import com.mkulesh.onpc.iscp.ConnectionIf;
import com.mkulesh.onpc.iscp.ISCPMessage;
import com.mkulesh.onpc.iscp.MessageChannel;
import com.mkulesh.onpc.iscp.State;
import com.mkulesh.onpc.iscp.messages.PowerStatusMsg;
import com.mkulesh.onpc.iscp.messages.ReceiverInformationMsg;
import com.mkulesh.onpc.utils.Logging;

import androidx.annotation.NonNull;

/**
 * The class performs receiver auto-power on startup
 **/
public class AutoPower implements MessageScriptIf
{
    public enum AutoPowerMode
    {
        POWER_ON,
        ALL_STANDBY
    }

    enum AllStandbyStep
    {
        NONE,
        ALL_STB_SEND,
        STB_SEND
    }

    private final AutoPowerMode autoPowerMode;
    private boolean done = false;
    AllStandbyStep allStandbyStep = AllStandbyStep.NONE;

    public AutoPower(final AutoPowerMode autoPowerMode)
    {
        this.autoPowerMode = autoPowerMode;
    }

    @Override
    public boolean isValid(ConnectionIf.ProtoType protoType)
    {
        return true;
    }

    @Override
    public boolean initialize(@NonNull final State state)
    {
        return isValid(state.protoType);
    }

    @Override
    public void start(@NonNull final State state, @NonNull MessageChannel channel)
    {
        Logging.info(this, "started script: " + autoPowerMode);
        done = false;
        allStandbyStep = AllStandbyStep.NONE;
    }

    @Override
    public void processMessage(@NonNull ISCPMessage msg, @NonNull final State state, @NonNull MessageChannel channel)
    {
        // Auto power-on once at first PowerStatusMsg
        if (msg instanceof PowerStatusMsg && !done)
        {
            final PowerStatusMsg pwrMsg = (PowerStatusMsg) msg;
            if (autoPowerMode == AutoPowerMode.POWER_ON && !state.isOn())
            {
                Logging.info(this, "request auto-power on startup");
                final PowerStatusMsg cmd = new PowerStatusMsg(state.getActiveZone(), PowerStatusMsg.PowerStatus.ON);
                channel.sendMessage(cmd.getCmdMsg());
                done = true;
            }
            else if (autoPowerMode == AutoPowerMode.ALL_STANDBY)
            {
                switch (allStandbyStep)
                {
                case NONE:
                {
                    Logging.info(this, "request all-standby on startup");
                    final PowerStatusMsg cmd = new PowerStatusMsg(
                            ReceiverInformationMsg.DEFAULT_ACTIVE_ZONE, PowerStatusMsg.PowerStatus.ALL_STB);
                    channel.sendMessage(cmd.getCmdMsg());
                    allStandbyStep = AllStandbyStep.ALL_STB_SEND;
                    break;
                }
                case ALL_STB_SEND:
                {
                    if (pwrMsg.getPowerStatus() == PowerStatusMsg.PowerStatus.STB ||
                            pwrMsg.getPowerStatus() == PowerStatusMsg.PowerStatus.ALL_STB)
                    {
                        done = true;
                    }
                    else
                    {
                        Logging.info(this, "request standby on startup");
                        final PowerStatusMsg cmd = new PowerStatusMsg(
                                ReceiverInformationMsg.DEFAULT_ACTIVE_ZONE, PowerStatusMsg.PowerStatus.STB);
                        channel.sendMessage(cmd.getCmdMsg());
                        allStandbyStep = AllStandbyStep.STB_SEND;
                    }
                    break;
                }
                case STB_SEND:
                    done = true;
                    break;
                }
                if (done)
                {
                    Logging.info(this, "close app after all-standby on startup");
                    System.exit(0);
                }
            }
        }
    }
}

File: lib/iscp/scripts/AutoPower.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */
// @dart=2.9
import 'package:flutter/services.dart';

import "../../utils/Logging.dart";
import "../ConnectionIf.dart";
import "../ISCPMessage.dart";
import "../MessageChannel.dart";
import "../State.dart";
import "../messages/PowerStatusMsg.dart";
import "../messages/ReceiverInformationMsg.dart";
import "MessageScriptIf.dart";

enum AutoPowerMode
{
    POWER_ON,
    ALL_STANDBY
}

enum _AllStandbyStep
{
    NONE,
    ALL_STB_SEND,
    STB_SEND
}

//
// The class performs receiver auto-power on startup
//
class AutoPower implements MessageScriptIf
{
    final AutoPowerMode autoPowerMode;
    bool _done = false;
    _AllStandbyStep _allStandbyStep = _AllStandbyStep.NONE;

    AutoPower(this.autoPowerMode);

    @override
    bool isValid(ProtoType protoType)
    => true;

    @override
    void initialize(final String data)
    {
        // nothing to do
    }

    @override
    void start(final State state, MessageChannel channel)
    {
        Logging.info(this, "started script: " + autoPowerMode.toString());
        _done = false;
        _allStandbyStep = _AllStandbyStep.NONE;
    }

    @override
    void processMessage(ISCPMessage msg, final State state, MessageChannel channel)
    {
        // Auto power-on once at first PowerStatusMsg
        if (msg is PowerStatusMsg && !_done)
        {
            if (autoPowerMode == AutoPowerMode.POWER_ON && !state.isOn)
            {
                Logging.info(this, "request auto-power on startup");
                final PowerStatusMsg cmd = PowerStatusMsg.output(state.getActiveZone, PowerStatus.ON);
                channel.sendMessage(cmd.getCmdMsg());
                _done = true;
            }
            else if (autoPowerMode == AutoPowerMode.ALL_STANDBY)
            {
                switch (_allStandbyStep)
                {
                case _AllStandbyStep.NONE:
                    Logging.info(this, "request all-standby on startup");
                    final PowerStatusMsg cmd = PowerStatusMsg.output(
                      ReceiverInformationMsg.DEFAULT_ACTIVE_ZONE, PowerStatus.ALL_STB);
                    channel.sendMessage(cmd.getCmdMsg());
                    _allStandbyStep = _AllStandbyStep.ALL_STB_SEND;
                    break;
                case _AllStandbyStep.ALL_STB_SEND:
                    if (msg.getValue.key == PowerStatus.STB || msg.getValue.key == PowerStatus.ALL_STB)
                    {
                        _done = true;
                    }
                    else
                    {
                        Logging.info(this, "request standby on startup");
                        final PowerStatusMsg cmd = PowerStatusMsg.output(
                            ReceiverInformationMsg.DEFAULT_ACTIVE_ZONE, PowerStatus.STB);
                        channel.sendMessage(cmd.getCmdMsg());
                        _allStandbyStep = _AllStandbyStep.STB_SEND;
                    }
                    break;
                case _AllStandbyStep.STB_SEND:
                    _done = true;
                    break;
                }
                if (_done)
                {
                    Logging.info(this, "close app after all-standby on startup");
                    SystemNavigator.pop();
                }
            }
        }
    }
}
